src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java

```diff
@@ -74,57 +74,66 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
             SegmentReader reader = (SegmentReader) FilterLeafReader.unwrap(context.reader());
             String directory = ((FSDirectory) FilterDirectory.unwrap(reader.directory())).getDirectory().toString();
 
-            /**
-             * In case of compound file, extension would be .hnswc otherwise .hnsw
-             */
-            String hnswFileExtension = reader.getSegmentInfo().info.getUseCompoundFile()
-                                               ? KNNCodecUtil.HNSW_COMPOUND_EXTENSION : KNNCodecUtil.HNSW_EXTENSION;
-            String hnswSuffix = knnQuery.getField() + hnswFileExtension;
-            List<String> hnswFiles = reader.getSegmentInfo().files().stream()
-                    .filter(fileName -> fileName.endsWith(hnswSuffix))
+            FieldInfo fieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
+
+            if (fieldInfo == null) {
+                logger.debug("[KNN] Field info not found for {}:{}", knnQuery.getField(),
+                        reader.getSegmentName());
+                return null;
+            }
+
+            KNNEngine knnEngine = KNNEngine.getEngine(fieldInfo.getAttribute(KNNConstants.KNN_ENGINE));
+            SpaceType spaceType = SpaceType.getSpace(fieldInfo.getAttribute(KNNConstants.SPACE_TYPE));
+            logger.debug("[KNN] knnEngine for " + knnQuery.getField() + ": " + knnEngine.getName());
+
+            String fileExtension = reader.getSegmentInfo().info.getUseCompoundFile()
+                                               ? knnEngine.getCompoundExtension() : knnEngine.getExtension();
+            String suffix = knnQuery.getField() + fileExtension;
+            List<String> engineFiles = reader.getSegmentInfo().files().stream()
+                    .filter(fileName -> fileName.endsWith(suffix))
                     .collect(Collectors.toList());
 
-            if(hnswFiles.isEmpty()) {
-                logger.debug("[KNN] No hnsw index found for field {} for segment {}",
+            if(engineFiles.isEmpty()) {
+                logger.debug("[KNN] No engine index found for field {} for segment {}",
                         knnQuery.getField(), reader.getSegmentName());
                 return null;
             }
 
             FieldInfo queryFieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
```

Comment:
This is similar to line 77. can we reuse it?

Reply:
You're right, I will fix this